app/service/learning_svc.py

```diff
@@ -35,43 +36,60 @@ async def build_model(self):
                         self.model.add(variables)
         self.model = set(self.model)
 
-    async def learn(self, facts, link, blob):
+    async def learn(self, facts, link, blob, operation=None):
         decoded_blob = b64decode(blob).decode('utf-8')
 
         found_facts = []
         for parser in self.parsers:
             try:
                 for fact in parser.parse(decoded_blob):
-                    await self._save_fact(link, facts, fact)
                     found_facts.append(fact)
             except Exception as e:
                 self.log.error(e)
         await self._update_scores(link, facts, increment=len(found_facts))
-        await self._build_relationships(link, found_facts)
+        await self._build_relationships(link, found_facts, operation)
 
     """ PRIVATE """
 
     @staticmethod
     async def _update_scores(link, facts, increment):
+        knowledge_svc_handle = BaseService.get_service('knowledge_svc')
         for uf in link.facts:
             for found_fact in facts:
                 if found_fact.unique == uf.unique:
                     found_fact.score += increment
+                    knowledge_svc_handle.update_fact(dict(trait=found_fact.trait, value=found_fact.value,
+                                                          source=found_fact.source), dict(score=found_fact.score))
                     break
 
     @staticmethod
-    async def _save_fact(link, facts, fact):
-        if all(fact.trait) and not any(f.trait == fact.trait and f.value == fact.value for f in facts):
+    async def _save_fact(link, facts, fact, operation=None):
+        if all(fact.trait) and not any(f.trait == fact.trait and f.value == fact.value and f.source == link.id
+                                       for f in facts):
             fact.collected_by = link.paw
             fact.technique_id = link.ability.technique_id
+            fact.links = [link]
+            fact.relationships = []
+            fact.source_type = OriginType.LEARNED.name
+            fact.source = operation.id if operation else link.id
+            knowledge_svc_handle = BaseService.get_service('knowledge_svc')
+            await knowledge_svc_handle.add_fact(fact)
             link.facts.append(fact)
 
-    async def _build_relationships(self, link, facts):
+    async def _build_relationships(self, link, facts, operation=None):
+        facts_covered = []
         for relationship in self.model:
```

Comment:
Whats the model here?

Reply:
The model here is whatever configuration has been loaded into the learning parser. It's why the learning parser is able to work 'dynamically', though the value of each individual model you could theoretically input varies.